pkg/featureflags/featureflags.go

```diff
@@ -0,0 +1,181 @@
+package featureflags
+
+import (
+	"context"
+	"fmt"
+	"slices"
+
+	"connectrpc.com/connect"
+
+	typesv1 "github.com/grafana/pyroscope/api/gen/proto/go/types/v1"
+)
+
+const (
+	V2StorageLayer          = "v2StorageLayer"
+	PyroscopeRuler          = "pyroscopeRuler"
+	PyroscopeRulerFunctions = "pyroscopeRulerFunctions"
+	UTF8LabelNames          = "utf8LabelNames"
+)
+
+type Scope string
+
+const (
+	ScopePusherV1  Scope = "pusherV1"  // this feature flag is relevant for the pusher.v1 API.
+	ScopeQuerierV1 Scope = "querierV1" // this feature flag is relevant for the querier.v1 API.
+)
+
+type featureDependency struct {
+	requires  []string // the feature cannot work without this other feature flag enabled.
+	conflicts []string // the feature cannot work with this other feature flag enabled.
+}
```

Comment:
I'm wondering if this is necessary.

I think that feature flags (capabilities) should just reflect the _actual_ configuration in use. E.g., if the "ruler" is enabled in the config/overrides, it should be visible in "feature flags", regardless of the logic that allows enabling the ruler.

If a parameter (config/override) depends on others, it should not be possible to set it to "impossible" value – such a configuration would be invalid. For example, we should never enable "ruler" or "read path symbolication" if v2 is not enabled.